ocaml/xapi/message_forwarding.ml

```diff
@@ -6464,4 +6472,183 @@ functor
               ~base_build_id ~base_version ~base_release ~to_version ~to_release
         )
     end
+
+    module Tracing = struct
+      let do_trace_op_on ~__context ~self ~local_fn ~client_fn ~db_fn =
+        let localhost = Helpers.get_localhost ~__context in
+        if Helpers.is_pool_master ~__context ~host:localhost then (
+          let hosts =
+            match Db.Tracing.get_hosts ~__context ~self with
+            | [] ->
+                Db.Host.get_all ~__context
+            | hosts ->
+                hosts
+          in
+          let hosts, local = List.partition (( <> ) localhost) hosts in
+          List.iter
+            (fun host -> do_op_on ~__context ~local_fn ~host client_fn)
+            hosts ;
+          List.iter (fun _ -> local_fn ~__context) local ;
+          db_fn ()
+        ) else
+          local_fn ~__context ;
+        ()
+
+      let create ~__context ~name_label ~hosts ~enabled ~tags ~endpoints
+          ~components ~filters ~processors =
+        let local_fn =
+          Local.Tracing.create ~name_label ~hosts ~enabled ~tags ~endpoints
+            ~components ~filters ~processors
+        in
+        let client_fn session_id rpc =
+          Client.Tracing.create ~rpc ~session_id ~name_label ~hosts ~enabled
+            ~tags ~endpoints ~components ~filters ~processors
+        in
+        let ref = Ref.make () in
+        let uuid = Uuidx.to_string (Uuidx.make ()) in
+        let localhost = Helpers.get_localhost ~__context in
+        if Helpers.is_pool_master ~__context ~host:localhost then (
+          let hosts =
+            match hosts with [] -> Db.Host.get_all ~__context | hosts -> hosts
+          in
+          let hosts, local = List.partition (( <> ) localhost) hosts in
+          List.iter
+            (fun host -> do_op_on ~__context ~local_fn ~host client_fn)
+            hosts ;
+          List.iter (fun _ -> local_fn ~__context) local ;
```

Comment:
Can't this be done using `do_trace_op_on`? There;s repeated logic here

Reply:
do_trace_op_on relies on one existing, it would be possible to create one then do the function with an empty db_fn

Reply:
Resolved [2eab70b](https://github.com/xapi-project/xen-api/pull/4966/commits/2eab70b666d23cd0747c1fc823356685c1b86e93)